konrad/convection.py

```diff
@@ -237,64 +328,45 @@ def calculate_convective_top(self, T_rad, T_con, p, timestep=0.1, lim=0.2):
-        # Convective heating must be positive somewhere
-        if np.any(convective_heating > lim):
-            # NOTE: `np.argmax` returns the first occurrence of the maximum value.
-            # In this example, the index of the first `True` value,
-            # corresponding to the convective top, is returned.
-            # The convective top corresponds to the first near zero / negative
-            # convective heating rate above the region which is being heated
-            # (the region which has a positive convective heating rate).
-            positive_i = int(np.argmax(convective_heating > lim))
-            contop_i = int(np.argmax(
-                convective_heating[positive_i:] < lim)) + positive_i
-
-            # Create auxiliary arrays storing the Qr, T and p values above and
-            # below the threshold value. These arrays are used as input for the
-            # interpolation in the next step.
-            heat_array = np.array([convective_heating[contop_i-1],
-                                   convective_heating[contop_i]])
-            p_array = np.array([p[contop_i-1], p[contop_i]])
-            T_array = np.array([T_con[contop_i-1], T_con[contop_i]])
-            index_array = np.array([contop_i-1, contop_i])
-
-            # Interpolate the pressure value to where the convective heating rate
-            # equals `lim`.
-            contop_index = interp1d(heat_array, index_array)(lim)
-            contop_p = interp1d(heat_array, p_array)(lim)
-            contop_T = interp1d(heat_array, T_array)(lim)
+        self.create_variable('convective_heating_rate', convective_heating)
 
-        else:
-            convective_heating = np.zeros(p.shape)
-            contop_index = np.nan
-            contop_p = np.nan
-            contop_T = np.nan
+        if np.any(convective_heating > lim):  # if there is convective heating
+            # find the values of pressure and temperature at the convective top
+            contop_p = interp_variable(p, convective_heating, lim)
+            contop_T = interp_variable(T_con, convective_heating, lim)
+            contop_index = interp_variable(
+                np.arange(0, p.shape[0]), convective_heating, lim
+            )
 
-        self.create_variable('convective_heating_rate', convective_heating)
-        self.create_variable('convective_top_plev', np.array([contop_p]))
-        self.create_variable('convective_top_temperature', np.array([contop_T]))
-        self.create_variable('convective_top_index', np.array([contop_index]))
+        else:  # if there is no convective heating
+            contop_index, contop_p, contop_T = np.nan, np.nan, np.nan
+
+        for name, value in [('convective_top_plev', contop_p),
+                            ('convective_top_temperature', contop_T),
+                            ('convective_top_index', contop_index),
+                            ]:
+            self.create_variable(name, np.array([value]))
 
         return
```

Comment:
I know this has been present before, but I think we should either return the convective top or maybe calls the method something like ``update_convective_top`` to make clear, that it updates the state of the object? The same is true for ``calculate_convective_top_height``